### DISCUSSION TITLE
Generating access token for another API than the account API?

### DISCUSSION THREAD

**Discussion by petertiedemann** [keycloak/keycloak#29175] [Q&A]
I work with an application consisting of the following:
A: An application frontend (SPA)
B: A backend API that stores data against OIDC identities (including roles and permissions)
C: Any external OIDC provider (Keycloak for the scope of this question)

A & B are part of the same application (so there are no trust issues besides A being an SPA).

B does *not* need to call any Keycloak API's, it only needs the subject and email from the token presented to it.

B expects a token that has audience set to a value specific to the API (or Client Id). For our purposes we can actually use the id token (because the actual authorization is done by our own API, we just need proof of identity), but libraries like oidc-client-ts don't handle refresh of id token, so we are a bit cautious about doing that in general as it causes expiry of the id token if access tokens has longer expiry than id tokens.

Comparing to Okta, they offer two kind of servers, the "Org" server, that provides access tokens for Okta's own API and "Authorization" servers that will issue a client/API specific access token. Interestingly Okta's **Org** server hard-codes the token lifetimes at 60 minutes, which might be to avoid the refresh issue mentioned above.

I am basically wondering what the Keycloak equivalent of the Okta "Authorization" server is?

**Comment by codespearhead** [comment]
I've never used Okta, but if you're asking how Keycloak emulates the existence of distinct authorization servers in a single Keycloak instance, then it's through [realms](https://www.keycloak.org/docs/latest/server_admin/index.html#configuring-realms):

> A realm is a space where you manage objects, including users, applications, roles, and groups. A user belongs to and logs into a realm. One Keycloak deployment can define, store, and manage as many realms as there is space for in the database.

**Reply by petertiedemann** [reply]
@codespearhead Thanks for replying, but I don't think that is quite what I am asking.

My concern is that the access token issued by Keycloak by default is meant to also be *consumed* by keycloak. It does not make sense for me to pass it to *my* API B (a resource server) as its granting access to a resource on another resource server (keycloak account resources). I am looking for a way for it to issue an access token authorizing frontend A to access API B on behalf the user.

Without this I either end up using the ID token to get access to the API (which is actually fine, but is not expected by libraries like oidc-client-ts etc).

Okta considers this an OAuth scenario that requires setting up an authorization server, but for example Entra ID is perfectly happy issuing such access tokens.

For anyone interested, this Okta Help Center article explain how they see it: https://support.okta.com/help/s/article/Signature-Validation-Failed-on-Access-Token?language=en_US

**Reply by codespearhead** [reply]
I suspect the main issue is translating the requirements from Okta-specific terms to OIDC-compliant terms.

> I am looking for a way for it to issue an access token authorizing frontend A to access API B on behalf the user

You seem to be describing the [Authorization Code Flow with PKCE](https://quarkus.io/guides/security-oidc-code-flow-authentication#:~:text=The%20Authorization%20Code%20Flow%20mechanism): the frontend is the public client whereas the backend API is the confidential client.

As for the second paragraph, it's by design because [access tokens are the ones meant to be used for authorization, not ID tokens](https://learn.microsoft.com/en-us/entra/identity-platform/id-tokens#:~:text=ID%20tokens%20shouldn't%20be%20used%20for%20authorization%20purposes).

**Reply by petertiedemann** [reply]
> You're describing the [Authorization Code Flow with PKCE](https://quarkus.io/guides/security-oidc-code-flow-authentication#:~:text=The%20Authorization%20Code%20Flow%20mechanism): the frontend is a public client, whereas the backend API is a confidential client.

Yes, that is the flow we are using (sorry I should have included that in the first post), except we don't access the user info endpoint (no use for it) and we **don't use sessions**.

> As for the second paragraph, it's by design because [access tokens are the ones meant to be used for authorization, not ID tokens](https://learn.microsoft.com/en-us/entra/identity-platform/id-tokens#:~:text=ID%20tokens%20shouldn't%20be%20used%20for%20authorization%20purposes).

Unsure about what paragraph you mean, but we do realize the id token is not intended for authorization. Our backend API handles its own authorization (because it holds users/role/permissions internally), it just needs proof of identity (authentication), which is why the id token actually can work for us, but it is something we are only doing for Okta Org (and now possibly Keycloak).

Because we don't rely on sessions, we are providing a token to the backend API in each request. With Entra this is the access token resulting from the login process, but with Okta Org server and Keycloak (with default configuration at least) we don't have such an access token.

Looking at the client settings in Keycloak, I strongly suspect the behavior I am interested in relates to these flags, but I honestly don't understand the description: 

![image](https://github.com/keycloak/keycloak/assets/1265354/3b89034f-28e3-4078-9402-159ff4b471df)
